polyply/src/apply_links.py

```diff
@@ -247,212 +162,152 @@ def _check_relative_order(resids, orders):
+    link_node_to_resid = {}
+    for resid, link_node in match.items():
+        for node in res_link.nodes[link_node]['graph']:
+            link_node_to_resid[node] = resid
+    return link_node_to_resid
 
-    Returns
-    ----------
-    tuple
-         list of nx.Graph
-             fragments of link in residue space
-         list of int
-             indices of link nodes
+def match_link_and_residue_atoms(meta_molecule, link, link_to_resid):
     """
-    paths = _orders_to_paths(meta_molecule, orders, node)
-    graphs = []
-    for path in paths:
-        graph = nx.Graph()
-        for idx, _node in enumerate(path[:-1]):
-            if _node != path[idx+1]:
-                graph.add_edge(_node, path[idx+1])
-        graphs.append(graph)
+    Given a meta_molecule a link and a correspondance of the link
+    nodes and those in the meta_molecule, establish a correspondance
+    between the link nodes and those in the higher resolution
+    molecule of the meta_molecule. Note that the meta_molecule needs
+    to have the "block" attribute discribing which atoms a single
+    meta_molecule node described.
 
-    return graphs, paths
-
-
-def is_subgraph(graph1, graph2):
-    """
-    Checks if graph1 is subgraph isomorphic to graph1.
-
-    Parameters
-    ----------
-    graph1: :class:`networkx.Graph`
-    graph2: :class:`networkx.Graph`
-
-    Returns
-    ----------
-    bool
-    """
-    for node in graph2.nodes:
-        if not node in graph1.nodes:
-           return False
-
-    for edge in graph2.edges:
-        if not graph1.has_edge(edge[0], edge[1]):
-           return False
-
-    return True
-
-
-def _get_link_resnames(link):
-    """
-    Get the resnames of a `link` directly from the node attributes
-    of the link atoms. It is not safe enough to just take the
-    link name, because it is often undefined or empty. Note that
-    a link name can also be a `vermouth.molecule.Choice` object.
-
-    Parameters
-    ----------
-    link: :class:`vermouth.molecule.Link`
+    Parameters:
+    -----------
+    meta_molecule: `:class:polyply.src.meta_molecule.MetaMolecule`
+    link:          `:class:vermouth.molecule.Link`
+    link_to_resid:  dict
+        correspondance dict of link nodes to meta_molecule nodes
 
-    Returns
-    ----------
-    set
-      all unique resnames of a molecule
+    Returns:
+    --------
+    dict
+        correspondance dict of link nodes to atoms in the
+        the meta_molecule.molecule attribute
     """
-    res_names = list(nx.get_node_attributes(link, "resname").values())
-    out_resnames = set()
+    link_to_mol = {}
+    for node in link.nodes:
+        meta_mol_key = link_to_resid[node]
+        block = meta_molecule.nodes[meta_mol_key]["graph"]
+        attrs = link.nodes[node]
+        attrs.update({'resid': meta_mol_key + 1})
+        ignore = ['order', 'charge_group', 'replace']
+        matchs = list(find_atoms(block, ignore=ignore, **attrs))
```

Comment:
I need to look into this more. That will be subject to a new PR, because it means we are starting to check a selection of edges described by a link and potentially links modify edges. So it's a bit of dangerous territory. At the moment the definition is very clear: A link always applies if the LCIS problem at the residue yields a match and the atoms are found within the two residues. That is much more clear cut than using partial edges for matching. However, having said that they is value in this, because it means you can batch up links even more. However, I want to have more test cases to decide if it is worth it or if you get into trouble.

Reply:
Also it only concerns interactions with more than 2 atoms.

Reply:
let's go first with just matching the atom-names. I think this is more clear and get's us a long way. We can always fall back to do something more fancy if we need to.